skyportal/facility_apis/ztf.py

```diff
@@ -130,19 +130,98 @@ def _build_forced_payload(self, request):
             'jdend': end_jd,
-            'ra': request.obj.ra,
-            'dec': request.obj.dec,
+            'ra': None,
+            'dec': None,
         }
 
-        return target
+        ra, dec = None, None
+        position = request.payload.get("position", "Source (object's coordinates)")
+        if "Flux weighted centroid" in position:
+            # here we query the database to get the all the photometry rows for the object that:
+            # 1. belong to an instrument with data stream(s)
+            # 2. have a non-null flux value
+            # 3. have a non-null flux error value
+            # 4. have coordinates (ra and dec)
+            # 5. do not have "fp" in the origin column
+            with DBSession() as session:
+                photometry = (
+                    session.scalars(
+                        sa.select(Photometry).where(
+                            sa.and_(
+                                Photometry.obj_id == request.obj.id,
+                                ~Photometry.origin.ilike('%fp%'),
+                            )
+                        )
+                    )
+                ).all()
+                additional_constraints = []
+                if "alert photometry only" in position:
+                    additional_constraints.append(lambda p: p and len(p.streams) > 0)
+                if "ZTF alert photometry only" in position:
+                    # grab the id of the ZTF instrument (one of ZTF or CFH12k)
+                    ztf_instrument_id = session.scalar(
+                        sa.select(Instrument.id).where(
+                            sa.or_(
+                                Instrument.name == "ZTF",
+                                Instrument.name == "CFH12k",
+                            )
+                        )
+                    )
+                    if ztf_instrument_id is not None:
+                        additional_constraints.append(
+                            lambda p: p and p.instrument_id == ztf_instrument_id
+                        )
+                    else:
+                        error = "Could not find the ZTF instrument in the database."
+                        return target, error
+                photometry = [
+                    (p.ra, p.dec, p.flux, p.fluxerr)
+                    for p in photometry
+                    if not np.isnan(p.flux)
+                    and not np.isnan(p.fluxerr)
+                    and not np.isnan(p.ra)
+                    and not np.isnan(p.dec)
+                    and len(p.streams) > 0
+                    and all(constraint(p) for constraint in additional_constraints)
+                ]
+                if len(photometry) > 0:
+                    # calculate the weighted centroid based on the object's photometry
+                    ras, decs, flux, fluxerr = np.array(photometry).T
+                    snr = flux / fluxerr
+                    ra, dec = np.sum(ras * snr) / np.sum(snr), np.sum(
```

Comment:
abs(snr)?

Reply:
Ah yes you are right, plus I said I would use that too, yikes.

Also wondering if I should put a cut on SNR to remove detections of low significance

Reply:
Ah yeah abs(snr) > 3 at least.

Reply:
added